File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-4.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode[]} lists
     * @return {ListNode}
     */
    mergeKLists(lists) {
        if(!lists || lists.length === 0) return null
        return this.divide(lists, 0, lists.length - 1);
    }

    divide(lists, l, r){
        if(l > r) return null
        if(l === r) return lists[l];

        const mid = Math.floor(l+ (r - l) / 2);
        const left = this.divide(lists, l, mid);
        const right = this.divide(lists, mid + 1, r);
        return this.conquer(left, right)
    }
    conquer(l1, l2){
        const dummy = new ListNode(0);
        let curr = dummy;

        while(l1 && l2){
            if(l1.val < l2.val){
                curr.next = l1;
                l1 = l1.next
            } else {
                curr.next = l2;
                l2 = l2.next;
            }
            curr = curr.next;
        }
        curr.next = l1 ? l1 : l2;
        return dummy.next;
    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-4.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int total = 0;
        int left =0;
        int right = n-1;
        int leftMax = height[left];
        int rightMax = height[right];

        while(left < right){

            if(leftMax < rightMax){
                left++;
                leftMax = Math.max(leftMax, height[left]);
                total += leftMax - height[left];
            } else {
                right--;
                rightMax = Math.max(rightMax, height[right]);
                total += rightMax - height[right];
            }
        }
        return total;
    }
}
